core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,251 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+import com.mesosphere.usi.models._
+
+object SchedulerLogic {
+  private def assertValidTransition(oldSpec: PodSpec, newSpec: PodSpec): Unit = {
+    if ((oldSpec.goal == Goal.Terminal) && (newSpec.goal == Goal.Running))
+      throw new IllegalStateException(s"Illegal state transition Terminal to Running for podId ${oldSpec.id}")
+  }
+
+  def taskIdsFor(pod: PodSpec): Seq[TaskId] = {
+    // ignoring momentarily that these might be different
+    Seq(pod.id.value)
+  }
+
+  def handlePodSpecUpdate(podSpecs: Map[PodId, PodSpec], podRecords: Map[PodId, PodRecord])
+                         (podId: PodId, newState: Option[PodSpec], responseBuilder: ResponseBuilder): ResponseBuilder = {
+    var b = responseBuilder
+
+    newState match {
+      case None =>
+        b = b.withChangeMessage(PodStatusChange(podId, None))
+        if (podRecords.contains(podId)) {
+          b = b.withChangeMessage(PodRecordChange(podId, None))
+        }
+
+      case Some(podSpec) =>
+
+        podSpecs.get(podId).foreach(assertValidTransition(_, podSpec))
+
+        if (podSpec.goal == Goal.Terminal) {
+          taskIdsFor(podSpec).foreach { taskId =>
+            b = b.withMesosCall(Mesos.Call.Kill(taskId))
+          }
+        }
+    }
+    b
+  }
+
+
+  def matchOffer(podSpecs: Map[PodId, PodSpec])(offer: Mesos.Event.Offer, pendingLaunch: Set[PodId], responseBuilder: ResponseBuilder): ResponseBuilder = {
```

Comment:
I like that this is a pure function. What is the motivation for using the builder pattern?

Reply:
We discussed this in a meeting